authprox/service.go

```diff
@@ -0,0 +1,322 @@
+package authprox
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	bolt "github.com/coreos/bbolt"
+	"github.com/dedis/cothority"
+	"github.com/dedis/kyber"
+	"github.com/dedis/kyber/share"
+	"github.com/dedis/kyber/sign/dss"
+	"github.com/dedis/kyber/suites"
+	"github.com/dedis/onet"
+	"github.com/dedis/onet/log"
+	"github.com/dedis/onet/network"
+)
+
+// ServiceName is the name of the Authentication Proxy service.
+const ServiceName = "AuthProx"
+
+var authProxID onet.ServiceID
+
+type service struct {
+	*onet.ServiceProcessor
+	ctx        context.Context
+	validators map[string]Validator
+	db         *bolt.DB
+	bucket     []byte
+}
+
+func init() {
+	var err error
+	authProxID, err = onet.RegisterNewService(ServiceName, newService)
+	if err != nil {
+		log.ErrFatal(err, "could not register")
+	}
+	network.RegisterMessages(
+		&EnrollRequest{}, &EnrollResponse{},
+		&SignatureRequest{}, &SignatureResponse{},
+		&EnrollmentsRequest{}, &EnrollmentsResponse{}, &EnrollmentInfo{},
+		&ti{}, &dssConfig{},
+	)
+}
+
+// A Validator is able to check the provided authInfo with respect to
+// the thrid-party authentication system. Return the user-id and the extra_data
+// associated with this auth info, or error.
+type Validator interface {
+	FindClaim(issuer string, authInfo []byte) (claim string, extraData string, err error)
+}
+
+func (s *service) registerValidator(t string, v Validator) {
+	if _, ok := s.validators[t]; ok {
+		panic("cannot re-register a validator")
+	}
+	s.validators[t] = v
+}
+
+func (s *service) validator(in string) (Validator, error) {
+	v, ok := s.validators[in]
+	if !ok {
+		return nil, fmt.Errorf("unknown auth type %v", in)
+	}
+	return v, nil
+}
+
+// ti is a struct to encode/decode a pair of type and issuer.
+type ti struct {
+	T, I string
+}
+
+type dssConfig struct {
+	Secret       kyber.Scalar
+	Participants []kyber.Point
+	LongPri      share.PriShare
+	LongPubs     []kyber.Point
+}
+
+func (s *service) find(typ, issuer string) (*dssConfig, error) {
+	k, err := network.Marshal(&ti{T: typ, I: issuer})
+	if err != nil {
+		return nil, err
+	}
+	var buf []byte
+	err = s.db.View(func(tx *bolt.Tx) error {
+		v := tx.Bucket(s.bucket).Get(k)
+		if v == nil {
+			return errors.New("claim not found")
+		}
+		// make a copy before leaving the tx
+		buf = append(buf, v...)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, x, err := network.Unmarshal(buf, cothority.Suite)
+	if err != nil {
+		return nil, err
+	}
+
+	var out *dssConfig
+	var ok bool
+	if out, ok = x.(*dssConfig); !ok {
+		return nil, fmt.Errorf("unexpected type %T", x)
+	}
+	return out, nil
+}
+
+func (s *service) Enrollments(req *EnrollmentsRequest) (*EnrollmentsResponse, error) {
```

Comment:
It is a private type, which is why lint did not catch it. I think I agree that Service entry points need to be documented, but am wondering now a bit about pub/priv on the type. (I'm happy it is private, but pondering if this will hurt us somehow?)